Document steps to support Docker-based HAProxy

fsat opened the pull request:
- Add installation instruction for Docker-based HAProxy.
- Add instructions for customised HAProxy reload script location.
- Add instructions for generate ConductR HAProxy configuration bundle for Docker-based HAProxy and customised HAProxy reload script location

fsat committed:
Document steps to support Docker-based HAProxy

- Add installation instruction for Docker-based HAProxy.
- Add instructions for customised HAProxy reload script location.
- Add instructions for generate ConductR HAProxy configuration bundle for Docker-based HAProxy and customised HAProxy reload script location

huntc commented on src/main/play-doc/operation/Install.md:
````diff
@@ -255,19 +259,112 @@ On some Debian distributions you may need to add a dedicated Personal Package Ar
+  bind :65535
+  mode http
+  monitor-uri /test
+```
````
I don't think that monitor code should be there...

edwardcallahan replied:
We certainly need to comment on the opening. What is it for? I should like to know more about this before allowing it.

huntc commented on src/main/play-doc/operation/Install.md:
````diff
@@ -255,19 +259,112 @@ On some Debian distributions you may need to add a dedicated Personal Package Ar
+Service ready.
+</body></html>
+* Closing connection 0
+```
````
I'm unsure about this monitor endpoint - why is that needed?

fsat replied:
Otherwise there will be no way to check if the HAProxy itself is running fine within the container.

I've done this when setting up the HAProxy within our DCOS cluster, I find it useful to have, and so I'd thought it would be useful to include it.

I'm happy to remove it though.

huntc replied:
I’d prefer to not expose endpoints unless they’re essential.

> On 22 Aug 2016, at 6:56 PM, Felix Satyaputra notifications@github.com wrote:
> 
> In src/main/play-doc/operation/Install.md https://github.com/typesafehub/conductr-doc/pull/251#discussion_r75639725:
> 
> > +> GET /test HTTP/1.1
> > +> Host: 10.0.7.118:65535
> > +> User-Agent: curl/7.48.0
> > +> Accept: _/_
> > +>
> > +\* HTTP 1.0, assume close after body
> > +< HTTP/1.0 200 OK
> > +< Cache-Control: no-cache
> > +< Connection: close
> > +< Content-Type: text/html
> > +<
> > +<html><body><h1>200 OK</h1>
> > +Service ready.
> > +</body></html>
> > +\* Closing connection 0
> > +```
> > Otherwise there will be no way to check if the HAProxy itself is running fine within the container.
> 
> I've done this when setting up the HAProxy within our DCOS cluster, I find it useful to have, and so I'd thought it would be useful to include it.
> 
> I'm happy to remove it though.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub https://github.com/typesafehub/conductr-doc/pull/251/files/ddfec2bedb2a9ba858b26571f8bd142b11db6079#r75639725, or mute the thread https://github.com/notifications/unsubscribe-auth/AAqabUniq3x7wprNTDAj9cqKrLiy14EFks5qiWREgaJpZM4Jpn7U.

edwardcallahan replied:
Don't remove it. Move it to a troubleshooting or as part of verification (if it doesn't verify) section.

huntc commented:
Looking good, just a comment on the need for a monitoring endpoint. Should be necessary.

@edwardcallahan Can you please review also?

fsat commented:
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;1fGGzQTg&#x2F;312-write-doco-for-docker-based-haproxy-installation"><img src="https:&#x2F;&#x2F;github.trello.services&#x2F;images&#x2F;trello-icon.png" width="12" height="12"> Write doco for Docker based HAProxy installation</a>

markusjura commented on src/main/play-doc/operation/Install.md:
```diff
@@ -255,19 +259,112 @@ On some Debian distributions you may need to add a dedicated Personal Package Ar
+
+#### Preparing HAProxy config
+
+Setup basic HAProxy configuration for the Docker based HAProxy. Ensure the `conductr-agent` user which runs the ConductR-HAProxy has ownership to the HAProxy config file.
```
`Ensure that the` instead of `Ensure the`.

markusjura commented on src/main/play-doc/operation/Install.md:
```diff
@@ -255,19 +259,112 @@ On some Debian distributions you may need to add a dedicated Personal Package Ar
+
+The proxy configuration located at `/etc/haproxy/haproxy.cfg` is mounted within the container on `/usr/local/etc/haproxy/haproxy.cfg`.
+
+The container exposes the port `9000`, `9999`, and `65535` to expose default port for ConductR HTTP-based endpoints, Visualizer bundle, and HAProxy test endpoint. Additional ports for your endpoints can be exposed through the `-p` option as required.
```
I'd say: "The container exposes the ports `9000`, `9999` and `65535` to the host to make the defaults ports of the ConductR HTTP-based endpoints, the Visualizer bundle and the HAProxy test endpoint available."

Note that the port `65535` can be removed here in case you remove the `frontend monitor` configuration in the `haproxy.cfg` as well, as Chris suggested in https://github.com/typesafehub/conductr-doc/pull/251/files#r75637541.

edwardcallahan replied:
Ah, this is why I might want to 'go native'.

In light of
http://stackoverflow.com/questions/28717464/docker-expose-all-ports-or-range-of-ports-from-7000-to-8000#comment51070632_28721643
Maybe we suggest using small ranges, 9000-9050, to simplify things without so many threads to wait for

markusjura commented on src/main/play-doc/operation/Install.md:
````diff
@@ -276,10 +373,17 @@ We will limit the bundle's sudo privileges to running `/usr/bin/reloadHAProxy.sh
 ```
 
-On RHEL and CentOS it may also be neccessary to [disable default requiretty](https://bugzilla.redhat.com/show_bug.cgi?id=1020147) for the `conductr-agent` user in `sudoers`.
+#### Using HAProxy reload script from a customised location
````
`customized` instead of `customised` (We generally use american english in our docs)

markusjura commented on src/main/play-doc/operation/Install.md:
```diff
@@ -276,10 +373,17 @@ We will limit the bundle's sudo privileges to running `/usr/bin/reloadHAProxy.sh
-On RHEL and CentOS it may also be neccessary to [disable default requiretty](https://bugzilla.redhat.com/show_bug.cgi?id=1020147) for the `conductr-agent` user in `sudoers`.
+#### Using HAProxy reload script from a customised location
+
+HAProxy reload script location can be customised. ConductR-HAProxy will install its reload script in this location upon startup.
```
`customized` instead of `customised` (We generally use american english in our docs)

markusjura commented on src/main/play-doc/operation/Install.md:
````diff
@@ -300,17 +404,182 @@ Append the `haproxy` role to the default `web` role as follows:
+export HAPROXY_RELOAD_SCRIPT_SOURCE="$CONFIG_DIR/reloadHAProxy.sh"
+```
+
+The `CONDUCTR_HAPROXY_CONFIG_OVERRIDE` and `HAPROXY_RELOAD_SCRIPT_SOURCE` supplies the customised HAProxy configuration and reload script respectively.
````
`customized` instead of `customised`.

markusjura commented:
Minor comments, otherwise LGTM.

edwardcallahan commented on src/main/play-doc/operation/Install.md:
```diff
@@ -233,7 +233,11 @@ _Perform each step in this section on all nodes: `172.17.0.1`, `172.17.0.2` and
 
-We will be using `HAProxy`. Install HAProxy version 1.5 or newer.
+We will be using `HAProxy` version 1.5 or newer.
+
```
It seems I have the choice of installing HAProxy Docker or Native. If so, why might I prefer one v the other and which is recommended.

edwardcallahan replied:
Suggested text:

ConductR-HAProxy requires `HAProxy` version 1.5 or newer. This can be satisfied by installing HAProxy natively, such as using a package manager, or by running the HAProxy Docker image. Native HAProxy installation requires less configuration but may not be available such as when using CoreOS on nodes.

edwardcallahan commented on src/main/play-doc/operation/Install.md:
````diff
@@ -255,19 +259,112 @@ On some Debian distributions you may need to add a dedicated Personal Package Ar
 [172.17.0.1]$ sudo chown conductr-agent:conductr-agent /etc/haproxy/haproxy.cfg
 ```
 
+Populate the `/etc/haproxy/haproxy.cfg` file with the following configuration.
````
Can we make this easier for operators to get at? Pop a copy on the downloads/conductr bucket namespace for example. 
Populate the `/etc/haproxy/haproxy.cfg` [file](https://downloads.typesafe.com/conductr/haproxy.cfg) with ...

edwardcallahan commented on src/main/play-doc/operation/Install.md:
```diff
@@ -255,19 +259,112 @@ On some Debian distributions you may need to add a dedicated Personal Package Ar
+The container exposes the port `9000`, `9999`, and `65535` to expose default port for ConductR HTTP-based endpoints, Visualizer bundle, and HAProxy test endpoint. Additional ports for your endpoints can be exposed through the `-p` option as required.
+
+The ports exposed by the container is bound to the `10.0.7.118` address as it's the interface where external traffic is expected to arrive from. Substitute the `10.0.7.118` address with the correct address in your environment.
+
```
something like 
 where requests will arrive from 
or 
 the service delivery interface

'external traffic'  is not correct enough here for me

edwardcallahan commented on src/main/play-doc/operation/Install.md:
````diff
@@ -830,4 +1372,4 @@ NAME           HOST       USER  STATE  ID
 visualizer-1   10.0.3.75  root    R    6e68f055d1f5715ad3ff19172fa5efaf_0f6e119c-9288-425a-89e3-36379dcaccda
 ```
 
````
Should we curl an endpoint from `conduct services` or the likes?
IOW, as an operator, where's my dmarc interface? The service might be returning 200's, but if it's not reachable, it's not verified.

edwardcallahan commented on src/main/play-doc/operation/Install.md:
```diff
@@ -233,7 +233,11 @@ _Perform each step in this section on all nodes: `172.17.0.1`, `172.17.0.2` and
 
-We will be using `HAProxy`. Install HAProxy version 1.5 or newer.
+We will be using `HAProxy` version 1.5 or newer.
+
```
Ah, with coreOS I must use Docker because I don't have apt/yum, no?

fsat committed:
Update docker run instruction to mount directory where haproxy.cfg is contained

This will ensure updates to haproxy.cfg to be made visible within the container which runs the HAProxy
